### DISCUSSION TITLE
How to get PPQ (pulse per quarter note) of midi file currently playing?

### DISCUSSION THREAD

**Discussion by denybear** [FluidSynth/fluidsynth#1112] [Q&A] [answered]
Hi,
Is there any function to get PPQ of the midi file that is currently playing?

I have tried 2 approaches:

1. use BPM and last tick information:

`

	/* keep running until the transport stops */
	while (1)
	{

		// load soundfont and midi file
		if (fluid_is_soundfont("/usr/share/sounds/sf2/FluidR3_GM.sf2")) {
			fluid_synth_sfload(synth, "/usr/share/sounds/sf2/FluidR3_GM.sf2", 1);
		}
		if (fluid_is_midifile("./ff7.mid")) {
			fluid_player_add(player, "./ff7.mid");
			printf ("file loaded\n");
		}

		// play the midi files, if any
		fluid_player_play(player);

		// get last tick
		int total_ticks;
		total_ticks = fluid_player_get_total_ticks(player);
		// get bpm
		int bpm;
		bpm = fluid_player_get_bpm(player);
		// compute ppq
		float ppq;
		ppq = total_ticks / bpm;
		printf ("total ticks: %d - bpm : %d - ppq: %f\n", total_ticks, bpm, ppq);

		// wait for playback termination
		fluid_player_join(player);
	}

`

But this returns BPM=120 (default) and total_ticks = 0

2. I have also tried to get `player->division` value, but it is 0 in the main()

How can I get PPQ of file that is currently playing?
Thanks

**Comment by denybear** [comment]
Ok, my above program does not work for 2 reasons:

- seems that calls such as `fluid_player_get_total_ticks ()` and `fluid_player_get_bpm ()` don't work unless you are in a callback (such as midi event callback, or midi tick callback)
- My program is wrong anyway:
```
	// get last tick
	int total_ticks;
	total_ticks = fluid_player_get_total_ticks(player);
	// get bpm
	int bpm;
	bpm = fluid_player_get_bpm(player);
	// compute ppq
	float ppq;
	ppq = total_ticks / bpm;
	printf ("total ticks: %d - bpm : %d - ppq: %f\n", total_ticks, bpm, ppq);

```
This only works if the duration of the midi file is 1 minute... which is never the case!


Unfortunately there is no call to get PPQ of MIDI file that is currently playing; 2 possibilities from there:

- either write a custom function in fluidsynth : takes only 2 lines of code to get PPQ (it is stored in `player->division`), but you end up with a custom fluidsynth library. However this works fine, I've done it.
- or load MIDI file to memory, read PPQ from memory, and use `fluid_player_add_mem ()` instead of `fluid_player_add ()`. I'm planning to try this soon.

**Reply by derselbst** [reply]
> takes only 2 lines of code to get PPQ (it is stored in player->division), but you end up with a custom fluidsynth library

You're welcome to propose a getter function for `player->divison`

**Reply by denybear** [reply]
@derselbst 
Happy to do that; is there any specific process to propose code to fluidsynth team, or should I just commit in the sourcecode?
Thanks

**Reply by derselbst** [reply]
https://github.com/FluidSynth/fluidsynth/blob/master/CONTRIBUTING.md#contributing

**Comment by Fruchtzwerg94** [comment, marked as answer]
See my PR #1220 which should solve the issue.
